### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by toanju**
a lot of refactoring to support future vlan, vxlan and l3 improvements

### CODE REVIEW

File: src/netlink/cnetlink.cpp [hunk 1]
```diff
@@ -16,31 +16,18 @@
...
-  enum link_type lt = LT_UNKNOWN;
-  for (const auto &n : lt2names) {
-    assert(lt != LT_MAX);
-    kind2lt.emplace(n, lt);
-    lt = static_cast<enum link_type>(lt + 1);
-  }
+cnetlink::cnetlink(std::shared_ptr<tap_manager> tap_man)
+    : swi(nullptr), thread(this), caches(NL_MAX_CACHE, nullptr),
+      tap_man(tap_man), bridge(nullptr), nl_proc_max(10), running(false),
+      rfd_scheduled(false), l3(new nl_l3(tap_man, this)) {
```
Comment (DFritzsche) [inline src/netlink/cnetlink.cpp]: where is the 'delete' for nl_l3?

File: src/netlink/netlink-utils.hpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <cassert>
+#include <cinttypes>
+#include <deque>
```
Comment (DFritzsche) [inline src/netlink/netlink-utils.hpp]: I think deque and assert are not used in this file

File: src/Makefile.am [hunk 3]
```diff
@@ -42,10 +42,13 @@ SOURCES_NETLINK = \
 	netlink/ctapdev.hpp \
 	netlink/nbi_impl.cpp \
 	netlink/nbi_impl.hpp \
+	netlink/netlink-utils.cpp \
+	netlink/netlink-utils.hpp \
 	netlink/nl_bridge.cpp \
 	netlink/nl_bridge.hpp \
-	netlink/nl_l3.hpp \
+	netlink/nl_hasing.hpp \
```
[the remark this answers is not in the document]
Comment (toanju) [reply src/Makefile.am]: hm, thought I fixed it here as well. Thanks!

File: src/netlink/cnetlink.cpp (hunk 1 above)
Comment (toanju) [reply src/netlink/cnetlink.cpp]: the shared_ptr will do this

File: src/netlink/netlink-utils.hpp (hunk 2 above)
Comment (toanju) [reply src/netlink/netlink-utils.hpp]: will check

File: src/Makefile.am (hunk 3 above)
[the remark this answers is not in the document]
Comment (toanju) [reply src/Makefile.am]: For compilation only this is not for interest. It would not be able to build a dist archive.

**Pull request merged**